src/version.c

```diff
@@ -10,31 +10,54 @@
 #include <funcapi.h>
 #include <fmgr.h>
 #include <storage/fd.h>
+#include <utils/timestamp.h>
 
 #include "fmgr.h"
 #include "compat.h"
 #include "gitcommit.h"
 #include "version.h"
 #include "config.h"
 
-#define STR_EXPAND(x) #x
-#define STR(x) STR_EXPAND(x)
-
-static const char *git_commit = STR(EXT_GIT_COMMIT);
+/* Export the strings to that we can read them using strings(1). We add a
+ * prefix so that we can easily find it using grep(1) */
+static const char commit_hash[] = "commit-hash:" EXT_GIT_COMMIT_HASH;
+static const char commit_tag[] = "commit-tag:" EXT_GIT_COMMIT_TAG;
+static const char commit_time[] = "commit-time:" EXT_GIT_COMMIT_TIME;
 
 TS_FUNCTION_INFO_V1(ts_get_git_commit);
 
 Datum
 ts_get_git_commit(PG_FUNCTION_ARGS)
 {
-	size_t var_size = VARHDRSZ + strlen(git_commit);
-	text *version_text = (text *) palloc(var_size);
+	TupleDesc tupdesc;
+	HeapTuple tuple;
+	Datum values[3];
+	bool nulls[3] = { false };
+	const char *const tag_ptr = strchr(commit_tag, ':');
```

Comment:
I don't understand why we first create a const char string for each commit information with their unique prefix, just to remove the prefix here.

If the idea is of the prefixes is to be able to find them in the binary (without calling this function), why not just directly use `EXT_GIT_COMMIT_TAG` here without having to strip the prefix that was just added above?

Reply:
Mainly because I want to ensure that the static variable is referenced so that it is not removed by an optimizing linker.

Reply:
Isn't it better to use a compiler pragma/attribute on the static variable instead to make it explicit that this variable should not be used.

The intention of this unnecessary parsing is non-obvious and it is easy for someone in the future to not understand its purpose and thus removing this code.

Reply:
Pragmas are usually not as portable, but I can switch.

Reply:
Yes, I realize that. But we use them for other things already, typically by defining a portable macro for the platforms we support. Usually, Windows is the odd-one out as GCC and Clang are often compatible.

Reply:
Added a new annotation and use that instead.